### ISSUE TITLE
Expose the rules list

### ISSUE THREAD

**Issue by mgol** [eslint/eslint#7021]
**The version of ESLint you are using.**
3.4.0

**The problem you want to solve.**
Since ESLint adds new rules from time to time and I want to make sure I make a decision about every core rule if I want to enable it or not, I've been [reading the `lib/rules/` contents](https://github.com/mgol/eslint-config-mgol/blob/65c4837cf076d2424ef2c0d744d07bb70ec2eea6/scripts/find-new-rules.js#L14) and fail if any of the rule names is not present in the config. This turned out to be a popular feature so it got extracted [into a separate package](https://www.npmjs.com/package/eslint-find-new-rules) and then [into a more general one](https://www.npmjs.com/package/eslint-find-new-rules). All of them apply the same hack of [reading the `lib/rules/` contents](https://github.com/sarbbottam/eslint-find-rules/blob/da18e90ebe3b2b84e226fead1d887543ad580c3b/src/lib/rule-finder.js#L74).

It would be good to have it available in the API and not require hacking into ESLint in such a way.

**Your take on the correct solution to problem.**
I want to have a list of all rule names available via the public API. Perhaps via something like:
```js
const rules= require('eslint').getRulesList();
```
(not sure if that's the best place but it doesn't fit for me to any of the current sub-APIs at http://eslint.org/docs/developer-guide/nodejs-api)

I'd expect the `rules` to be an array of rule definitions where rule definition is more or less what's in the `meta` object exported by the rule (+ its name), e.g.:
```js
{
    name: 'comma-dangle',

    docs: {
        description: "require or disallow trailing commas",
        category: "Stylistic Issues",
        recommended: false
    },

    fixable: "code",

    schema: [
        {
            enum: ["always", "always-multiline", "only-multiline", "never"]
        }
    ]
},
```

Returning the whole object instead of just rule names would allow to do more analysis for the API, e.g. it'd allow to not require specifying deprecated rules: https://github.com/sarbbottam/eslint-find-rules/issues/172

If you're afraid the whole meta objects is too much information, at least the following would help a lot:
```js
[
    {
        name: 'comma-dangle',
        deprecated: false,
    },
    {
        name: 'no-native-reassign',
        deprecated: true,
    },
]
```

**Comment by jfmengels** [comment]
This feature would be really nice :+1: 

> If you're afraid the whole meta objects is too much information, at least the following would help a lot:

I think that the whole meta object would be nice for plenty of different tools or uses, and it would be a shame to only limit it to a few fields. But I don't know how much memory it uses and if that causes problems.

**Comment by vitorbal** [comment]
Thanks for the detailed description. I agree this is a great idea, so much so that we already have it on our radar. Check it out: https://github.com/eslint/eslint/issues/6525
`getRules` should be exactly what you're proposing, right?

**Comment by jfmengels** [comment]
Sounds like it to me, thanks @vitorbal

**Comment by mgol** [comment]
@vitorbal Yes, that looks good! Somehow I haven't found this one when I was searching for similar issues. 😄 

Should we close this one as a duplicate then?

**Comment by platinumazure** [comment]
Closing as duplicate- thanks for the excellent suggestion!
